libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,613 @@
+     * index of the quasar in the current candidate array.*/
+    return qso - ncand_before;
+}
+
+/* Calculates the total ionization fraction of the box.
+ */
+static double
+gas_ionization_fraction(void)
+{
+    int64_t n_ionized_tot = 0, n_gas_tot = 0;
+    int i, n_ionized = 0;
+    #pragma omp parallel for reduction(+:n_ionized)
+    for (i = 0; i < PartManager->NumPart; i++){
+        if (P[i].Type == 0 && P[i].Ionized == 1){
+            n_ionized ++;
+        }
+    }
+    /* Get total ionization fraction: note this is only the current gas particles.
+     * Particles that become stars are not counted.*/
+    sumup_large_ints(1, &n_ionized, &n_ionized_tot);
+    sumup_large_ints(1, &SlotsManager->info[0].size, &n_gas_tot);
+    return (double) n_ionized_tot / (double) n_gas_tot;
+}
+
+/* Do the ionization for a single particle, marking it and adding the heat.
+ * No locking is performed so ensure the particle is not being edited in parallel.
+ * Returns 1 if ionization was done, 0 otherwise.*/
+static int
+ionize_single_particle(int other)
+{
+    /* Only ionize things once*/
+    if(P[other].Ionized != 0)
+        return 0;
+
+    /* Mark it ionized*/
+    P[other].Ionized = 1;
+
+    /* Heat the particle*/
+
+    /* Number of helium atoms per g in the particle*/
+    double nheperg = (1 - HYDROGEN_MASSFRAC) / (PROTONMASS * HEMASS);
+    /* Total heating per unit mass ergs/g for the particle*/
+    double deltau = qso_inst_heating * nheperg;
+    double uu_in_cgs = All.UnitEnergy_in_cgs / All.UnitMass_in_g;
+
+    /* Conversion factor between internal energy and entropy.*/
+    double entropytou = pow(SPH_EOMDensity(other) * All.cf.a3inv, GAMMA_MINUS1) / GAMMA_MINUS1;
+    /* Convert to entropy in internal units*/
+    SPHP(other).Entropy += deltau / uu_in_cgs / entropytou;
+    return 1;
+}
+
+struct QSOPriv {
+    /* Particle SpinLocks*/
+    struct SpinLocks * spin;
+    FOFGroups * fof;
+};
+#define QSO_GET_PRIV(tw) ((struct QSOPriv *) (tw->priv))
+
+/**
+ * Ionize and heat the particles
+ */
+static void
+ionize_ngbiter(TreeWalkQueryBase * I,
+        TreeWalkResultBase * O,
+        TreeWalkNgbIterBase * iter,
+        LocalTreeWalk * lv)
+{
+
+    if(iter->other == -1) {
+        /* Gas only*/
+        iter->mask = 1;
+        /* Bubble size*/
+        double bubble = gaussian_rng(QSOLightupParams.mean_bubble, sqrt(QSOLightupParams.var_bubble), I->ID);
+        iter->Hsml = bubble;
+        /* Don't care about gas HSML */
+        iter->symmetric = NGB_TREEFIND_ASYMMETRIC;
+        return;
+    }
+
+    int other = iter->other;
+
+    /* Only ionize gas*/
+    if(P[other].Type != 0)
+        return;
+
+    struct SpinLocks * spin = QSO_GET_PRIV(lv->tw)->spin;
+
+    lock_spinlock(other, spin);
+
+    int ionized = ionize_single_particle(other);
+
+    unlock_spinlock(other, spin);
+
+    if(!ionized)
+        return;
+
+    /* Add to the ionization counter*/
+    #pragma omp atomic
```

Comment:
Do we have a better place to store these variables, e.g. in the Iterator? I remember @sbird may have tried to eliminate thread contentions due to atomics?

Reply:
Excellent point!